spring-cloud-loadbalancer/src/main/java/org/springframework/cloud/client/loadbalancer/impl/context/RoundRobinLoadBalancer.java

```diff
@@ -0,0 +1,90 @@
+package org.springframework.cloud.client.loadbalancer.impl.context;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cloud.client.ServiceInstance;
+import org.springframework.cloud.client.loadbalancer.impl.scoped.ScopedDiscoveryClient;
+import org.springframework.cloud.client.loadbalancer.impl.support.LoadBalancerClientFactory;
+
+import static org.springframework.util.CollectionUtils.isEmpty;
+
+/**
+ * @author Spencer Gibb
+ */
+public class RoundRobinLoadBalancer implements LoadBalancer<ServiceInstance> {
+
+	private static final Log log = LogFactory.getLog(RoundRobinLoadBalancer.class);
+
+	private final LoadBalancerClientFactory clientFactory;
+	private final AtomicInteger nextServerCyclicCounter = new AtomicInteger(-1);
+
+	@Value("${loadbalancer.client.name}")
+	String serviceId;
+
+	public RoundRobinLoadBalancer(LoadBalancerClientFactory clientFactory) {
+		this.clientFactory = clientFactory;
+	}
+
+	@Override
+	public ChosenContext<ServiceInstance> choose(RequestContext requestContext) {
+		ScopedDiscoveryClient discoveryClient = this.clientFactory.getInstance(this.serviceId, ScopedDiscoveryClient.class);
+		List<ServiceInstance> instances = discoveryClient.getInstances();
+
+		if (isEmpty(instances)) {
+			log.warn("No servers available for service: " + this.serviceId);
+			return new DefaultChosenContext(null);
+		}
+
+		int nextServerIndex = incrementAndGetModulo(instances.size());
+
+		ServiceInstance instance = instances.get(nextServerIndex);
+
+		return new DefaultChosenContext(instance);
+	}
+
+	public static class DefaultChosenContext implements ChosenContext<ServiceInstance> {
```

Comment:
can things like this still operate package private? if we aren't exposing new methods, I'd prefer them hidden